### PR TITLE
Hex dma2

### PR DESCRIPTION
**Pull request by vgundlur**
1. Made Changes to Support both LUMA and chroma(NV12) data in single Buffer
2. Adjusted the Mock DMA transfer code and the hex_dma example to do DMA on Luma+Chorma
3. Currently we assumed that Chroma processing happens always after Luma processing, We are working on to avoid this Assumption
4. We have verified the UBWC compressed Luma+chroma (NV12) transfer with this PR
5. 10-bit format p010 works
6. Enhancements to support other formats like TP10, compressed 10-bit format is in progress
7. Currently We are Supporting RAW Format for 2D and 1D only, Higher dimensions for RAW will be supported Later

### CODE REVIEW

File: apps/hexagon_dma/pipeline.cpp [hunk 1]
```diff
@@ -5,33 +5,48 @@ using namespace Halide;
 class DmaPipeline : public Generator<DmaPipeline> {
 public:
     Input<Buffer<uint8_t>> input{"input", 2};
```
Comment (dsharletg) [inline apps/hexagon_dma/pipeline.cpp]: This also needs to be 3 dimensional?

File: apps/hexagon_dma/process.cpp [hunk 2]
```diff
@@ -39,38 +37,60 @@ int main(int argc, char **argv) {
                              reinterpret_cast<uint64_t>(memory_to_dma_from));
     input.set_device_dirty();
 
+    //Halide::Runtime::Buffer<uint8_t> output(width, height, 2);
```
Comment (dsharletg) [inline apps/hexagon_dma/process.cpp]: Remove or comment why this is removed.

File: apps/hexagon_dma/process.cpp [hunk 3]
```diff
@@ -39,38 +37,60 @@ int main(int argc, char **argv) {
...
+
     // In order to actually do a DMA transfer, we need to allocate a
     // DMA engine.
     void *dma_engine = nullptr;
     halide_hexagon_dma_allocate_engine(nullptr, &dma_engine);
 
     // We then need to prepare for copying to host. Attempting to copy
     // to host without doing this is an error.
-    halide_hexagon_dma_prepare_for_copy_to_host(nullptr, input, dma_engine);
+    halide_hexagon_dma_prepare_for_copy_to_host(nullptr, input, dma_engine, false, eDmaFmt_NV12,0);
```
Comment (dsharletg) [inline apps/hexagon_dma/process.cpp]: Space after last `,` (2x).

File: apps/hexagon_dma/process_p010.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        for(int y = 0; y < height_c; y++) {
+            for(int x = 0; x < width; x++) {
+                uint16_t correct = memory_to_dma_from[x + y*width + z*width*height] * 2;
+                if (correct != output(x, y, z)) {
+                    static int cnt = 0;
+                    printf("Mismatch at x=%d y=%d c=%d : %d != %d\n", x, y, z, correct, output(x, y, z));
+                    if(++cnt > 20) abort();
+                }
+            }
+         }
```
Comment (dsharletg) [inline apps/hexagon_dma/process_p010.cpp]: Indentation looks off by a space.

File: src/runtime/hexagon_dma.cpp [hunk 5]
```diff
@@ -121,6 +127,150 @@ static void desc_pool_free (void* user_context) {
...
+    stDmaTransferParm.pFrameBuf             = dev->buffer;
+    if (dev->is_write) {
+       stDmaTransferParm.eTransferType      = eDmaWrapper_L2ToDdr;
+    } else {
+       stDmaTransferParm.eTransferType      = eDmaWrapper_DdrToL2;
+    }
+    stDmaTransferParm.u16RoiX               = dev->offset_x + dst->dim[0].min;
+    stDmaTransferParm.u16RoiY               = dev->offset_y + dst->dim[1].min;
+
+    // DMA Driver Halves the Height and Y Offset so that only Half the ROI Size of Luma is transferred  for chroma
```
Comment (dsharletg) [inline src/runtime/hexagon_dma.cpp]: Something I'm still confused about: does the DMA hardware do anything with the information that these planes are (Y)UV data? i.e. does it do chroma upsampling/downsampling, or RGB <-> YUV conversion?

File: src/runtime/hexagon_dma.cpp [hunk 6]
```diff
@@ -121,6 +127,150 @@ static void desc_pool_free (void* user_context) {
...
+    uint8_t *dest = reinterpret_cast<uint8_t *>(dst->host);
+
+    // Copy from Locked Cache to a temp DDR buffer
+    // TODO: This should be removed once the cache locking is addressed inside Halide Pipeline
+    if (dest) {
+        const int pixelsize = dst->type.bytes();
+        const int linesize = roi_width * pixelsize;
+        uint8_t *cache_buf = reinterpret_cast<uint8_t *>(dev->cache_buf);
+        debug(user_context) << "copy cache: " << "pixelsize= " << pixelsize << " linesize= " << linesize << "\n";
+        int y, sy, dy;
```
Comment (dsharletg) [inline src/runtime/hexagon_dma.cpp]: I'd prefer it if these were declared "inline":

        for (int y = 0; y < roi_height; y++) {
            int sy = y * roi_stride * pixelsize;
            int dy = y * dst->dim[1].stride * pixelsize;
            memcpy(&dest[dy], &cache_buf[sy], linesize);
        }

The generated code should be the same, and it's easier to read/safer (IMO). Also note spacing changes.

File: src/runtime/hexagon_dma.cpp [hunk 7]
```diff
@@ -121,6 +127,150 @@ static void desc_pool_free (void* user_context) {
...
+    
+    debug(user_context) << "Hexagon: " << dev->dma_engine << " move\n" ;
+
+    nRet = nDmaWrapper_Move(dev->dma_engine);
+    if (nRet != QURT_EOK) {
+        debug(user_context) << "Hexagon: nDmaWrapper_Move error: " << nRet << "\n";
+        return halide_error_code_device_buffer_copy_failed;
+    }
+    
+   debug(user_context) << "Hexagon: " << dev->dma_engine << " wait\n" ;
```
Comment (dsharletg) [inline src/runtime/hexagon_dma.cpp]: indent a space

File: apps/hexagon_dma/pipeline.cpp [hunk 8]
```diff
@@ -4,34 +4,51 @@ using namespace Halide;
 
 class DmaPipeline : public Generator<DmaPipeline> {
 public:
-    Input<Buffer<uint8_t>> input{"input", 2};
-    Output<Buffer<uint8_t>> output{"output", 2};
+    Input<Buffer<uint8_t>> input{"input", 3};
+    Output<Buffer<uint8_t>> output_y{"output_y", 3};
```
[the remark this answers is not in the document]
Comment (vgundlur) [reply apps/hexagon_dma/pipeline.cpp]: Agree with you on this but we have done this to distinguish the luma transfers in Two DMA Formats: eDmaFmt_NV12_Y and eDmaFmt_NV12. We are using Two dimension buffer for eDmaFmt_NV12_Y and 3-D buffer for Luma in eDmaFmt_NV12

File: src/runtime/hexagon_dma.cpp (hunk 5 above)
Comment (vgundlur) [reply src/runtime/hexagon_dma.cpp]: DMA hardware does the down sampling automatically when we choose the format as eDmaFmt_NV12_UV

Comment (dsharletg) [reply src/runtime/hexagon_dma.cpp]: OK, so one would only use the YUV formats when one actually wants the DMA hardware to do a conversion, right? If just processing YUV data without conversion, we would just use the basic data types?

**Comment by dsharletg** [comment]
It looks like the last push pulled in a bunch of irrelevant commits...

File: apps/hexagon_dma/process.cpp [hunk 9]
```diff
@@ -46,23 +43,36 @@ int main(int argc, char **argv) {
...
-    // Validate that the algorithm did what we expect.
-    output.for_each_element([&](int x, int y) {
-        uint8_t correct = input_validation(x, y) * 2;
-        if (correct != output(x, y)) {
-            printf("Mismatch at %d %d: %d != %d\n", x, y, correct, output(x, y));
-            abort();
+    output.copy_to_host();
+    int c = 2;
+    const int plane_start = 0;
+    const int plane_end = c;
```
Comment (dsharletg) [inline apps/hexagon_dma/process.cpp]: c -> 2, or better yet, just substitute in plane_start and plane_end, I don't see why these need to be variables.

File: apps/hexagon_dma/process.cpp [hunk 10]
```diff
@@ -46,23 +43,36 @@ int main(int argc, char **argv) {
...
+    printf("plane start=%d end=%d\n", plane_start, plane_end);
+    for (int z = plane_start; z < plane_end; z++) {
```
Comment (dsharletg) [inline apps/hexagon_dma/process.cpp]: z -> c

File: apps/hexagon_dma/process.cpp [hunk 11]
```diff
@@ -46,23 +43,36 @@ int main(int argc, char **argv) {
...
     // We then need to prepare for copying to host. Attempting to copy
     // to host without doing this is an error.
-    halide_hexagon_dma_prepare_for_copy_to_host(nullptr, input, dma_engine);
+    // The Last parameter 0 indicate DMA Read
+    halide_hexagon_dma_prepare_for_copy_to_host(nullptr, input, dma_engine, false, eDmaFmt_NV12);
 
-    Halide::Runtime::Buffer<uint8_t> output(width, height);
+    Halide::Runtime::Buffer<uint8_t> output(width, height, 2);
+    Halide::Runtime::Buffer<uint8_t> output_y = output.cropped(2, 0, 1);    // Luma plane only
+    Halide::Runtime::Buffer<uint8_t> output_c = output.cropped(2, 1, 1).cropped(1, 0, (height/2));  // Chroma plane only, with reduced height
```
[the remark this answers is not in the document]
Comment (dsharletg) [reply apps/hexagon_dma/process.cpp]: It also seems like this allocates a buffer that is half a plane too big.

File: src/runtime/hexagon_dma.cpp [hunk 12]
```diff
@@ -121,6 +127,150 @@ static void desc_pool_free (void* user_context) {
...
+        << "Hexagon dev handle: buffer: " << dev->buffer << " dev offset (" << dev->offset_x << ", " << dev->offset_y << ") frame_width: " << dev->frame_width
+        << " frame_height: " << dev->frame_height << " frame_stride: " << dev->frame_stride << "\n";
+
+    int roi_stride = 0;
+    int roi_width = 0;
+    int roi_height = 0;
+
+    // Changing the Format to Chroma or LUMA based on dimension    
+    t_eDmaFmt currentFmt = dev->fmt;
+    if (dst->dimensions == 3) {
```
Comment (dsharletg) [inline src/runtime/hexagon_dma.cpp]: We've got to find another way to give the DMA engine the information it needs. I think the basic problem is that we're trying to do something here that Halide just doesn't understand: you have a single input buffer with a Y plane (W x H) and a UV plane (W/2 x H/2 x 2), and you're trying to read it in a Halide program as two separate buffers. This necessary because Halide doesn't understand buffers with channels of different dimensions, but the approach you are taking is also highly suspect.

I really, really wish we could focus on plain-old-data rather than formats with "built-in implications" like YUV 4:2:0.

Basically, the problem you've undertaken here really is not a DMA engine problem, it's a YUV 4:2:0 problem. I don't have an easy answer for this.

File: apps/hexagon_dma/process_p010.cpp [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    Halide::Runtime::Buffer<uint16_t> output_y = output.cropped(2, 0, 1);    // Luma plane only
+    Halide::Runtime::Buffer<uint16_t> output_c = output.cropped(2, 1, 1).cropped(1, 0, (height/2));  // Chroma plane only, with reduced height
+
+    int result = pipeline_p010(input, output_y, output_c);
+    if (result != 0) {
+        printf("pipeline failed! %d\n", result);
+    }
+
+    output.copy_to_host();
+    int c = 2;
```
Comment (dsharletg) [inline apps/hexagon_dma/process_p010.cpp]: Same comments as on the other test

File: apps/hexagon_dma/process.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (vgundlur) [reply apps/hexagon_dma/process.cpp]: We need to have One Output buffer (DMA Requirement) to hold both the Y and UV data, How to specify asymmetric plane size to meet the Width * Height * 3 / 2 requirement ?

File: apps/hexagon_dma/process.cpp (hunk 9 above)
Comment (vgundlur) [reply apps/hexagon_dma/process.cpp]: Agreed and corrected

File: apps/hexagon_dma/process.cpp (hunk 10 above)
Comment (vgundlur) [reply apps/hexagon_dma/process.cpp]: Agreed and corrected

File: apps/hexagon_dma/process_p010.cpp (hunk 13 above)
Comment (vgundlur) [reply apps/hexagon_dma/process_p010.cpp]: Agreed and corrected

File: src/runtime/hexagon_dma.cpp (hunk 12 above)
Comment (vgundlur) [reply src/runtime/hexagon_dma.cpp]: Agree that we have to find a different way to address this but currently this is the best solution. Once there is a Halide solution for YUV 420 we will adopt accordingly.

We are using W * H * 2 as an interim solution so that we can flatten the Halide Buffer to allow the DMA to function properly using one buffer.

File: apps/hexagon_dma/pipeline_err.cpp [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+#define DIM 3
+
+class DmaPipeline : public Generator<DmaPipeline> {
+public:
+    Input<Buffer<uint8_t>> input{"input", DIM};
+    Output<Buffer<uint8_t>> output{"output", DIM};
+
+    void generate() {
+        Var x{"x"}, y{"y"}, z("z");
```
Comment (dsharletg) [inline apps/hexagon_dma/pipeline_err.cpp]: z -> c

File: apps/hexagon_dma/process_nv12.cpp [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    // DMA engine.
+    void *dma_engine = nullptr;
+    halide_hexagon_dma_allocate_engine(nullptr, &dma_engine);
+
+    // We then need to prepare for copying to host. Attempting to copy
+    // to host without doing this is an error.
+    // The Last parameter 0 indicate DMA Read
+    halide_hexagon_dma_prepare_for_copy_to_host(nullptr, input, dma_engine, false, eDmaFmt_NV12);
+
+    Halide::Runtime::Buffer<uint8_t> output(width, height, 2);
```
Comment (dsharletg) [inline apps/hexagon_dma/process_nv12.cpp]: You can allocate a buffer of the correct size in a few ways. Allocate a regular buffer (via malloc, etc.), and then create Halide::Runtime::Buffers that point into that buffer, or, something like this is equivalent to what you have now:

    Buffer<uint8_t> output(width, height * 3 / 2);
    Buffer<uint8_t> output_y = output.cropped(1, 0, height);
    Buffer<uint8_t> output_uv = output.cropped(1, height, height / 2);

You can further make this correct by giving uv two channels:

    output_uv.embed(2, 0);
    output_uv.dim(2).extent = 2;
    output_uv.dim(2).stride = 1;
    output_uv.dim(0).extent /= 2;
    output_uv.dim(0).stride = 2;

I'm not completely sure about the last part, but conceptually it should be correct (you want to split the x dimension into an inner dimension with extent 2 and an outer dimension of extent width / 2).

**Comment by vgundlur** [comment]
Hi Dillon

Thank You for the Suggestions. We want to split the problem in to Two parts.


  1.  Avoid the Extra 0.5 W*H buffer allocation
     *   We are solving by making slight modifications to the solution you have suggested.
  2.  UV de-interleaving
     *   instead of de-interleaving in Halide Buffer, we are de-interleaving in the Algorithm section of pipeline code

We are doing this to be consistent with the DMA function.  Please see the commit : 10bad0139d46f11c9fa4531e680db13f5cc792b1 for the changes we have made.
Please share your comments.

Note: The commit above just have the front end changes and doesn’t include the backend changes(hexagon_dma.cpp) to make it functional.
           Pending your confirmation to this approach then we will make the changes to the backend.

Thanks,
Venu.

File: apps/hexagon_dma/process_nv12.cpp [hunk 16]
```diff
@@ -47,11 +47,19 @@ int main(int argc, char **argv) {
     // The Last parameter 0 indicate DMA Read
     halide_hexagon_dma_prepare_for_copy_to_host(nullptr, input, dma_engine, false, eDmaFmt_NV12);
 
-    Halide::Runtime::Buffer<uint8_t> output(width, height, 2);
+    //Halide::Runtime::Buffer<uint8_t> output(width, height, 2);
+    Halide::Runtime::Buffer<uint8_t> output(width, (3 * height) / 2);
+    output.embed(2, 0);
```
Comment (dsharletg) [inline apps/hexagon_dma/process_nv12.cpp]: This (and the following 3 lines) should be done to the chroma buffer only, after cropping it out of the outer buffer)

File: apps/hexagon_dma/pipeline_nv12.cpp [hunk 17]
```diff
@@ -15,12 +16,16 @@ class DmaPipeline : public Generator<DmaPipeline> {
...
         Func copy_y("copy_y");
         Func copy_uv("copy_uv");
+        Func uv_de_interleaved("uv_de_interleaved");
+        Func processed_uv("processed_uv");
 
-        copy_y(x, y, c) = input(x, y, c);
-        copy_uv(x, y, c) = input(x, y, c);
+        copy_y(x, y) = input_y(x, y);
+        copy_uv(x, y, c) = input_uv(x, y, c);
+        uv_de_interleaved(x, y, c) = copy_uv(x/2*2, y/2, c);
```
Comment (dsharletg) [inline apps/hexagon_dma/pipeline_nv12.cpp]: Add parens to make this more clear: `(x/2)*2`

That said, assuming we're building the "ideal" thing here (without backend changes yet, per your PR comment), this pipeline should look like this:

        copy_y(x, y) = input_y(x, y);
        copy_uv(x, y, c) = input_uv(x, y, c);
 
        output_y(x, y) = copy_y(x, y) * 2;
        output_uv(x, y, c) = copy_uv(x, y, c) * 2;

Because the DMA engine will need the UV buffer to be interleaved, you will need to do this in the schedule:

        // For internal pipeline buffers, we can just reorder the storage so c is innermost, matching the DMA engine expectations
        // (this is an interleaved buffer).
        copy_uv.reorder_storage(c, x, y)

        // For input/output buffers, we need to tweak the expected extents/strides:
        input_uv.dim(0).set_stride(2);
        input_uv.dim(2).set_stride(1).set_extent(2);
        output_uv.output_buffer().dim(0).set_stride(2);
        output_uv.output_buffer().dim(2).set_stride(1).set_extent(2);

File: apps/hexagon_dma/process_nv12.cpp [hunk 18]
```diff
@@ -47,11 +47,19 @@ int main(int argc, char **argv) {
...
+    output.raw_buffer()->dim[1].extent = height;
+    output.raw_buffer()->dim[2].extent = 2;
+    output.raw_buffer()->dim[2].stride = width * height;
+      
     Halide::Runtime::Buffer<uint8_t> output_y = output.cropped(2, 0, 1);    // Luma plane only
```
Comment (dsharletg) [inline apps/hexagon_dma/process_nv12.cpp]: This also doesn't quite look right. I think the whole buffer construction should look like this:

    Halide::Runtime::Buffer<uint8_t> output(width, (3 * height) / 2);

    // Extract the Y plane
    Halide::Runtime::Buffer<uint8_t> output_y = output.cropped(1, 0, height);

    // Extract the UV plane.
    Halide::Runtime::Buffer<uint8_t> output_uv = output.cropped(1, height, height/2);

    // Add a channels dimension. At this point, the buffer has dimensions width x height/2 x 1
    output_uv.embed(2, 0);

    // Rewrite the x dimension so the buffer now has dimensions width/2 x height/2 x 1, though the x dimension has stride 2
    output_uv.raw_buffer()->dim[0].extent = width / 2;
    output_uv.raw_buffer()->dim[0].stride = 2;

    // Verify that the height is what we expect.
    assert(output_uv.raw_buffer()->dim[1].extent == height / 2);

    // Make the channels dimension have extent 2, and stride 1 (pairs of UV values are stored adjacently in memory).
    output_uv.raw_buffer()->dim[2].stride = 1; 
    output_uv.raw_buffer()->dim[2].extent = 2;  
    
    // Now the buffer has dimensions width/2 x height/2 x 2, and the channels are stored interleaved.

Comment (dsharletg) [reply apps/hexagon_dma/process_nv12.cpp]: Is the only change from what is committed to this branch the pipeline you copy/pasted into this comment?

Comment (dsharletg) [reply apps/hexagon_dma/process_nv12.cpp]: It looks like the host code is still checking for a planar chroma buffer, but the pipeline you showed above outputs an interleaved buffer, so that seems like the source of a potential problem (though that should cause an assert to fail, not bad output).

For reference, I'm referring to line 75 of apps/hexagon_dma/process_nv12.cpp, which appears to be loading a plane (the stride of 'c' is width*height) from the reference input.
